contracts/vesting/src/contract.rs

```diff
@@ -0,0 +1,517 @@
+    ) -> Result<(), ContractError>;
+
+    fn transfer_token(
+        env: Env,
+        from: Address,
+        to: Address,
+        amount: i128,
+    ) -> Result<(), ContractError>;
+
+    fn burn(env: Env, sender: Address, amount: i128) -> Result<(), ContractError>;
+
+    fn mint(env: Env, sender: Address, to: Address, amount: i128);
+
+    // TODO: we will need these in the future, not needed for the most basic implementation right now
+    // TODO: replace the tuple `owner_spender: (Address, Address)` with how it is in `send_to_contract_from`
+    // fn increase_allowance(env: Env, owner_spender: (Address, Address), amount: i128);
+
+    // fn decrease_allowance(env: Env, owner_spender: (Address, Address), amount: i128);
+
+    // fn transfer_from(
+    //     env: Env,
+    //     owner_spender: (Address, Address),
+    //     to: Address,
+    //     amount: i128,
+    // ) -> Result<(), ContractError>;
+
+    // fn burn_from(
+    //     env: Env,
+    //     sender: Address,
+    //     owner: Address,
+    //     amount: i128,
+    // ) -> Result<(), ContractError>;
+
+    // fn send_to_contract_from(
+    //     env: Env,
+    //     sender: Address,
+    //     owner: Address,
+    //     contract: Address,
+    //     amount: i128,
+    // ) -> Result<(), ContractError>;
+
+    fn update_minter(env: Env, sender: Address, new_minter: Address);
+
+    fn update_minter_capacity(env: Env, sender: Address, new_capacity: u128);
+
+    fn query_balance(env: Env, address: Address) -> i128;
+
+    fn query_distribution_info(
+        env: Env,
+        address: Address,
+    ) -> Result<DistributionInfo, ContractError>;
+
+    fn query_token_info(env: Env) -> VestingTokenInfo;
+
+    fn query_minter(env: Env) -> MinterInfo;
+
+    fn query_vesting_total_supply(env: Env) -> i128;
+}
+
+#[contractimpl]
+impl VestingTrait for Vesting {
+    fn initialize(
+        env: Env,
+        admin: Address,
+        vesting_token: VestingTokenInfo,
+        vesting_balances: Vec<VestingBalance>,
+        minter_info: Option<MinterInfo>,
+        max_vesting_complexity: u32,
+    ) -> Result<(), ContractError> {
+        save_admin(&env, &admin);
+
+        if vesting_balances.is_empty() {
+            log!(
+                &env,
+                "Vesting: Initialize: At least one vesting schedule must be provided."
+            );
+            panic_with_error!(env, ContractError::MissingBalance);
+        }
+
+        let total_supply = create_vesting_accounts(&env, max_vesting_complexity, vesting_balances)?;
+        if let Some(mi) = minter_info {
+            let input_curve = Curve::Constant(mi.mint_capacity);
+
+            let capacity = input_curve.value(env.ledger().timestamp()) as i128;
+            if total_supply > capacity {
+                log!(&env, "Vesting: Initialize: total supply over the capacity");
+                panic_with_error!(env, ContractError::SupplyOverTheCap);
+            }
+            save_minter(&env, &mi);
```

Comment:
During initialize, add transaction of token's transfer from `admin` to the contract's address.